### ISSUE TITLE
Make all the Atoms immutable

### ISSUE THREAD

**Issue by ForNeVeR** [ForNeVeR/wpf-math#125]
At the beginning I've tried to do that as part as #123, but now I have a "bright" idea to land it to the master branch _before_ merging #123.

So, in #123, we've got ourselves into various troubles because our `Atom`s are mutable objects: we can't cache them, it's difficult to (safely) clone them etc.

I think we need to make all of our `Atom`s immutable (i.e. no writable or otherwise mutable properties). It will require some nontrivial changes to our code base, but I think it'll be worth.

Some of the methods' signatures will change, but as `Atom` is an `internal` class, almost none of our public APIs will change (and that's always a good thing).

That's a critical task for #123.

**Comment by B3zaleel** [comment]
We have to make them mutable to support the *\\atop*, *\\choose* and *\\over* commands.

**Comment by ForNeVeR** [comment]
I don't think so. Generally, in the cases you want to change something you just construct new changed copy of immutable data. In practical applications it tends to work very well.

Consider you want `\atop {ATOM}` to change the `ATOM` is some way. It should work like that:

1. Immutable `ATOM` is constructed by the parser.
2. `\atop` command wants to change `ATOM` in one way or another
3. `\atop` just takes the `ATOM` and creates a copy with changed data (e.g. new height value or such).

We, indeed, have some issues with creating immutable copies of our stuff, but all of them are solvable. Feel free to ask questions!
